### ISSUE TITLE
Unexpected behavior when kernel takes list of booleans

### ISSUE THREAD

**Issue by jezerjojo14** [NVIDIA/cuda-quantum#2262]
### Required prerequisites

- [x] Consult the [security policy](https://github.com/NVIDIA/cuda-quantum/security/policy). If reporting a security vulnerability, do not report the bug using this form. Use the process described in the policy to report the issue.
- [x] Make sure you've read the [documentation](https://nvidia.github.io/cuda-quantum/latest). Your issue may be addressed there.
- [x] Search the [issue tracker](https://github.com/NVIDIA/cuda-quantum/issues) to verify that this hasn't already been reported. +1 or comment there if it has.
- [ ] If possible, make a PR with a failing test to give us a starting point to work on!

### Describe the bug

I needed to send a list of $n$ booleans as a parameter to a kernel to specify which of $n$ qubits to apply $X$ gates on. The kernel does not throw any errors, but does not seem to be interpreting this list of booleans correctly. I'm not sure but I think it might be interpreting the list of booleans as a single integer?

### Steps to reproduce the bug

This is the kernel:

```
import cudaq
@cudaq.kernel
def circ_test(n: int, bools: list[bool]):
    q = cudaq.qvector(n)
    for j in range(n):
        b=bools[j]
        if b==False:
            x(q[j])

```

This is how it behaves for different examples:

1)
```
print(cudaq.draw(circ_test,5,[True,True,True,True,True]))
          
q0 : ─────
     ╭───╮
q1 : ┤ x ├
     ├───┤
q2 : ┤ x ├
     ├───┤
q3 : ┤ x ├
     ├───┤
q4 : ┤ x ├
     ╰───╯


```

2)
```
print(cudaq.draw(circ_test,5,[False,True,True,False,True]))

q0 : ─────
     ╭───╮
q1 : ┤ x ├
     ├───┤
q2 : ┤ x ├
     ├───┤
q3 : ┤ x ├
     ├───┤
q4 : ┤ x ├
     ╰───╯
```

3)

```
print(cudaq.draw(circ_test,5,[False,False,False,False,False]))

     ╭───╮
q0 : ┤ x ├
     ├───┤
q1 : ┤ x ├
     ├───┤
q2 : ┤ x ├
     ├───┤
q3 : ┤ x ├
     ├───┤
q4 : ┤ x ├
     ╰───╯
```

Here's a modified kernel:

```
import cudaq
@cudaq.kernel
def circ_test_2(n: int, bools: list[bool]):
    q = cudaq.qvector(n)
    for j in range(n):
        b=bools[j]
        if b==True:
            x(q[j])
```

1)
```
print(cudaq.draw(circ_test_2,5,[False,False,False,False,False]))
```
No output

2)
```
print(cudaq.draw(circ_test_2,5,[True,True,True,True,True]))
```
No output

3)
```
print(cudaq.draw(circ_test_2,5,[False,True,False,False,True]))
```
No output

4)
```
print(cudaq.draw(circ_test_2,5,[True,False,False,False,False]))

     ╭───╮
q0 : ┤ x ├
     ╰───╯
```

### Expected behavior

For the first kernel, I expect an $X$ gate to be applied on the $i$-th qubit if and only if the $i$-th element of the boolean list is False. For the second kernel I expect the same but for when the $i$-th element of the boolean list is True.

### Is this a regression? If it is, put the last known working version (or commit) here.

Not a regression

### Environment

- **CUDA Quantum version**: CUDA-Q Version 0.8.0 (https://github.com/NVIDIA/cuda-quantum 5f63a898a1c4af1affd293fcc86220b683436f57)
- **Python version**: 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0]
- **Operating system**: WSL 2 (Ubuntu) on Windows 11

**Comment by bebora** [comment]
Hi @jezerjojo14 , I tested some other variants of your code, changing the value being checked, the defined type for the list and the actual values in the list.
## Boolean list, boolean comparison
```python
import cudaq
@cudaq.kernel
def circ_test(n: int, bools: list[bool]):
    q = cudaq.qvector(n)
    for j in range(n):
        b=bools[j]
        if b==True:
            x(q[j])

print(cudaq.draw(circ_test,5,[False,True,False,True,False]))
```
Output:
Empty
If we were to call the kernel with a list of integers, it would throw an error:
```python
print(cudaq.draw(circ_test,5,[0,1,0,1,0]))
```
Output:
`RuntimeError: kernel argument's element type is 'bool' but argument provided is not (argument 1, element 0, value=0, type=<class 'int'>).`
## Boolean list, integer comparison
```python
import cudaq
@cudaq.kernel
def circ_test(n: int, bools: list[bool]):
    q = cudaq.qvector(n)
    for j in range(n):
        b=bools[j]
        if b==1:
            x(q[j])

print(cudaq.draw(circ_test,5,[False,True,False,True,False]))
```
Output:
Empty
## Integer list, boolean comparison
```python
import cudaq
@cudaq.kernel
def circ_test(n: int, bools: list[int]):
    q = cudaq.qvector(n)
    for j in range(n):
        b=bools[j]
        if b==True:
            x(q[j])

print(cudaq.draw(circ_test,5,[0,1,0,1,0]))
```
Output:
```
error: 'arith.cmpi' op requires all operands to have the same type
RuntimeError: Failure while executing pass pipeline.
```
## Integer list, integer comparison
```python
import cudaq
@cudaq.kernel
def circ_test(n: int, bools: list[int]):
    q = cudaq.qvector(n)
    for j in range(n):
        b=bools[j]
        if b==1:
            x(q[j])

print(cudaq.draw(circ_test,5,[0,1,0,1,0]))
```
Output:
```
q0 : ─────
     ╭───╮
q1 : ┤ x ├
     ╰───╯
q2 : ─────
     ╭───╮
q3 : ┤ x ├
     ╰───╯
```
If we were to call the kernel with a list of booleans, it would work too:
```python
print(cudaq.draw(circ_test,5,[False,True,False,True,False]))
```
Output:
```
q0 : ─────
     ╭───╮
q1 : ┤ x ├
     ╰───╯
q2 : ─────
     ╭───╮
q3 : ┤ x ├
     ╰───╯
```
## Takeaway
You can currently call your kernel passing a list of booleans as long as the kernel treats it as a list of integers. There is likely something wrong when using lists of booleans inside quantum kernels. A simplified version of your code where the kernel receives a `bool` instead of a `list[bool]` works fine, so the bool type is not problematic per se.

**Comment by jezerjojo14** [comment]
I'm curious as to what was going wrong in the first place but this is a great solution to the problem I was facing. Thanks @bebora 👍

**Comment by schweitzpgi** [comment]
Fixed with #2338
